contents/df_end.md

```diff
@@ -0,0 +1,224 @@
+grade_2020(2)
+) # hide
+""")
+```
+
+Continuing on this, we can also get only `names` for the first 2 rows:
+
+```jl
+@sco JDS.grades_indexing()
+```
+
+If we assume that all names in the table are unique, we can also write a function to obtain the grade for a person via their `name`.
+To do this, we convert the table back to one of Julia's basic data structures which is capable of creating a mappings, namely dictionaries:
+
+```jl
+@sc grade_2020("")
+```
+
+```jl
+sco("""
+grade_2020("Bob")
+""")
+```
+
+which works because `zip` loops through `df.name` and `df.grade_2020` at the same time like a zipper:
+
+```jl
+sco("""
+df = grades_2020()
+collect(zip(df.name, df.grade_2020))
+""")
+```
+
+However, converting a DataFrame to a Dictionary is only useful when the elements are unique.
+When that is not the case, it is time to `filter` (@sec:filter).
+
+## Filter and Subset {#sec:filter}
+
+Continuing from the earlier mentioned table.
+
+```jl
+sco("""
+without_caption_label( # hide
+grades_2020()
+) # hide
+""")
+```
+
+We can filter rows by using `filter(f::Function, df)`.
+This function is very similar to the function `filter(f::Function, V::Vector)` from Julia itself.
+Working with a function `f` for filtering can be a bit difficult to use in practice, but it is very powerful.
+As a simple example, we can create a function which checks whether it's input equals "Alice":
+
+```jl
+@sc JDS.equals_alice("")
+```
+
+```jl
+sco("""
+equals_alice("Bob")
+""")
+```
+
+```jl
+sco("""
+equals_alice("Alice")
+""")
+```
+
+With this function, we can now filter out all the rows for which `name` equals "Alice"
+
+```jl
+sco("""
+without_caption_label( # hide
+filter(:name => equals_alice, grades_2020())
+) # hide
+""")
+```
+
+Note that this doesn't only work for DataFrames, but also for vectors:
+
+```jl
+sco("""
+filter(equals_alice, ["Alice", "Bob", "Dave"])
+""")
+```
+
+We can make it a bit less verbose by using an anonymous function:
+
+```jl
+sco("""
+filter(n -> n == "Alice", ["Alice", "Bob", "Dave"])
+""")
+```
+
+or, for the table,
+
+```jl
+sco("""
+without_caption_label( # hide
+filter(:name => n -> n == "Alice", grades_2020())
+) # hide
+""")
+```
+
+This line can be read as "for each element in the column `:name`, let's call this thing `n`, check whether `n` equals Alice".
+For some people, this is still to verbose.
+Luckily, Julia has added a _partial function application_ of `==`.
+The details of these words are not important, only that you can use it via
+
+```jl
+sco("""
+s = "Workaround a bug in books" # hide
+without_caption_label( # hide
+filter(:name => ==("Alice"), grades_2020())
+) # hide
+""")
+```
+
+To get all the rows which are **not** Alice, `==` can be replaced by `!=` in all previous examples:
+
+```jl
+sco("""
+without_caption_label( # hide
+filter(:name => !=("Alice"), grades_2020())
+) # hide
+""")
+```
+
+Now, to show why these functions are so powerful, we can come up with a more complex filter.
```

Comment:
```suggestion
Now, to show why anonymous functions are so powerful, we can come up with a more complex filter.
```

Too much this/these/that etc....
What you think?

Reply:
Agreed. Thanks for the suggestions. I've tried to replace some.